### ISSUE TITLE
What is wrong with my model? It slows many times after switching from version 0.4.1 to 1.0

### ISSUE THREAD

**Issue by CoinCheung** [pytorch/pytorch#14456]
This is the definition of my model: 
```python
import torchvision
import torch
import torch.nn as nn
import torch.nn.functional as F



class Model(nn.Module):
    def __init__(self, in_dim, out_dim, *args, **kwargs):
        super(Model, self).__init__(*args, **kwargs)
        vgg16 = torchvision.models.vgg16()

        layers = []
        layers.append(nn.Conv2d(in_dim, 64, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.MaxPool2d(3, stride = 2, padding = 1))

        layers.append(nn.Conv2d(64, 128, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.Conv2d(128, 128, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.MaxPool2d(3, stride = 2, padding = 1))

        layers.append(nn.Conv2d(128, 256, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.Conv2d(256, 256, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.Conv2d(256, 256, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.MaxPool2d(3, stride = 2, padding = 1))

        layers.append(nn.Conv2d(256, 512, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.Conv2d(512, 512, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.Conv2d(512, 512, kernel_size = 3, stride = 1, padding = 1))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.MaxPool2d(3, stride = 1, padding = 1))

        layers.append(nn.Conv2d(512,
            512,
            kernel_size = 3,
            stride = 1,
            padding = 2,
            dilation = 2))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.Conv2d(512,
            512,
            kernel_size = 3,
            stride = 1,
            padding = 2,
            dilation = 2))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.Conv2d(512,
            512,
            kernel_size = 3,
            stride = 1,
            padding = 2,
            dilation = 2))
        layers.append(nn.ReLU(inplace = True))
        layers.append(nn.MaxPool2d(3, stride = 1, padding = 1))
        self.features = nn.Sequential(*layers)

        classifier = []
        classifier.append(nn.AvgPool2d(3, stride = 1, padding = 1))
        classifier.append(nn.Conv2d(512,
            1024,
            kernel_size = 3,
            stride = 1,
            padding = 12,
            dilation = 12))
        classifier.append(nn.ReLU(inplace = True))
        classifier.append(nn.Conv2d(1024, 1024, kernel_size = 1, stride = 1, padding = 0))
        classifier.append(nn.ReLU(inplace = True))
        classifier.append(nn.Dropout(p = 0.5))
        classifier.append(nn.Conv2d(1024, out_dim, kernel_size = 1))
        self.classifier = nn.Sequential(*classifier)

        self.init_weights()


    def forward(self, x):
        im = x
        x = self.features(x)
        x = self.classifier(x)
        return x

    def init_weights(self):
        vgg = torchvision.models.vgg16(pretrained = True)
        state_vgg = vgg.features.state_dict()
        self.features.load_state_dict(state_vgg)

        for ly in self.classifier.children():
            if isinstance(ly, nn.Conv2d):
                nn.init.kaiming_normal_(ly.weight, a=1)
                nn.init.constant_(ly.bias, 0)
```
And this is my test script: 
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import time
from model import Model


if __name__ == "__main__":

    net = Model(3, 21)
    net.train()
    net.cuda()
    net = nn.DataParallel(net)
    Loss = nn.CrossEntropyLoss(ignore_index = 255)
    Loss.cuda()
    optim = torch.optim.SGD(net.parameters(), lr = 1e-3, momentum = 0.9, weight_decay = 5e-4)

    st = time.time()
    scale = [0.5, 0.75, 1]
    loss_avg = []
    for i in range(10000):
        in_ten = torch.randn(70, 3, 224, 224)
        label = torch.randint(0, 21, [70, 1, 224, 224])
        in_ten = in_ten.cuda()
        label = label.cuda()
        label = torch.tensor(label).long().cuda()
        optim.zero_grad()
        H, W = in_ten.size()[2:]
        for sub_i, s in enumerate(scale):
            print(time.time() - st)
            h, w = int(H * s), int(W * s)
            in_ten_s = F.interpolate(in_ten, (h, w), mode = 'bilinear')
            out = net(in_ten_s)
            out = F.interpolate(out, [H, W], mode = 'bilinear')
            label = torch.squeeze(label)
            loss = Loss(out, label)
            loss.backward()
            loss = loss.detach().cpu().numpy()
            loss_avg.append(loss)
        optim.step()

        if i % 20 == 0 and not i == 0:
            ed = time.time()
            interval = ed - st
            st = ed
            loss_avg = sum(loss_avg) / len(loss_avg)
            print('iter: {}, time: {}, loss: {}'.format(i, interval, loss_avg))
            loss_avg = []
```

The problem is that I can run this program quite fast with pytorch 0.4.1, however, after switching to nightly version of `1.0.0.dev20181127`, it slows down a great deal. 
I install this nightly version with `pip install --user torch_nightly -f https://download.pytorch.org/whl/nightly/cu90/torch_nightly.html`, and I am working on ubuntu16.04 with python3.5. By the way, I use two 1080ti gpu to run this script.

After some trying, I got to find the problem is likely to be with the `Model` class, since after replacing it with some resnet50 (with last fc and avg_pool layers removed), it will run much faster, although it still takes some time to run this line: `loss = loss.detach().cpu().numpy()`

Hope I have made myself clear, and what is the cause of my program ? Did I misunderstand the phylosophy of writing pytorch programs ?

**Comment by fehiepsi** [comment]
We also observe the slowness when running `nn.Conv2d` on Travis. It is flaky: sometimes it is fast, sometimes it is slow (0.8s vs 0.008s). The flakiness happens for both python 2 and python 3.

I can't replicate the problem locally in my machine. It just happens on Travis. So it might be an installation issue. (we install PyTorch in Travis using the link `http://download.pytorch.org/whl/cpu/torch-1.0.0-cp27-cp27mu-linux_x86_64.whl`.)

**Comment by CoinCheung** [comment]
My problem is solved when I compile and install from the source.

**Comment by sshuair** [comment]
I got the same problem, much slower than 0.4.1, both pip and conda.

**Comment by AlexTaguchi** [comment]
Same problem here as well, around 3-4 times slower training on my 1080 Ti GPU with pytorch 1.0 than 0.4 (haven't tried installing from source)

**Comment by AlexTaguchi** [comment]
Here's a minimalized model that shows the behavior on my machine (Ubuntu 16, Python 3.5, NVIDIA GTX 1080 Ti):

```
class Model(nn.Module):

    # Neural Network Architecture
    def __init__(self, hidden_size):
        super().__init__()

        # self.dilation = [1, 1, 1, 1, 1, 1, 1]
        self.dilation = [1, 2, 4, 8, 16, 32, 64]

        # Build first layer of convolutional network
        self.layers = nn.ModuleList([nn.Conv1d(4, hidden_size, 3, padding=1)])
        self.layers.extend([nn.BatchNorm1d(hidden_size)])
        self.layers.extend([nn.ReLU()])

        # Build remaining convolutional network
        for n in self.dilation:
            self.layers.extend([nn.Conv1d(hidden_size, hidden_size, 3, padding=n, dilation=n)])
            self.layers.extend([nn.BatchNorm1d(hidden_size)])
            self.layers.extend([nn.ReLU()])

        # Final convolutional layer
        self.layers.extend([nn.Conv1d(hidden_size, 2, 3, padding=1, dilation=1)])

    # Neural Network Implementation
    def forward(self, in):

        output = in

        # Pass through convolutional layers
        for x in self.layers:
            output = x(output)
        output = torch.transpose(output, 1, 2)

        return output
```

Interestingly, in my case it seems the implementation of dilated convolutional layers is causing the huge difference in training speed for me. With Pytorch 0.4 (0.4.1-py35_cuda9.2.148_cudnn7.1.4_1 pytorch [cuda92]) the above model will train almost an order of magnitude faster than with Pytorch 1.0 (1.0.0-py3.5_cuda10.0.130_cudnn7.4.1_1 pytorch [cuda100]). On the other hand, when I change the commentation in the script above to the following (removing any convolutional dilation) the performance on Pytorch 0.4 is the same as on 1.0:
```
self.dilation = [1, 1, 1, 1, 1, 1, 1]
# self.dilation = [1, 2, 4, 8, 16, 32, 64]
```

Note: I haven't looked into whether or not the difference in CUDA version is affection anything (9.2 for Pytorch 0.4 and 10.0 for Pytorc 1.0)

Hope this is of some help

**Comment by CoinCheung** [comment]
Aye, in my case, there is also heavily usage of atrous convolution.

**Comment by sshuair** [comment]
@AlexTaguchi @CoinCheung I have `dilation` too.

**Comment by TinkeringCode** [comment]
+1 I have a similar problem with LSTMs.
My (small) LSTM model runs about 3x slower in PyTorch 1.0.0 installed from conda than in PyTorch 0.4.1 on the CPU.

**Comment by ezyang** [comment]
@TinkeringCode Can you try compiling from source and seeing if there is still a problem? We have fixed some possible problems on master.

**Comment by TillChima** [comment]
I am experiencing a similar behaviour with pytorch 1.0 instead of 0.4. 
My model also contains lots of Atrous / Dilated Convolutions.
Has anybody been able to fix this issue properly?

**Comment by knazeri** [comment]
@TillChima apparently, the problem comes from CuDNN 7.4 trying to find the best algorithm on a particular GPU, for specific sizes of convolution / dilation. Soumith's solution worked for me: https://github.com/pytorch/pytorch/issues/15054#issuecomment-450191923
If your input size does not change, you can enable CuDNN becnhmark mode:
`torch.backends.cudnn.benchmark=True`

**Comment by CoinCheung** [comment]
I should close this, since the  reason is found. Thanks for support :)
